### ISSUE TITLE
Dealing with two very different transaction classes

### ISSUE THREAD

**Issue by rbrunner7** [seraphis-migration/wallet3#7]
Monero transactions have changed considerably over the course of their history since 2014, with the introduction of RingCT, with the switch from Borromean ring signatures to Bulletproofs and very recently to Bulletproofs+.

Despite all these changes there always was and still is a single class `cryptonote::transaction`, as defined in the code [here](https://github.com/monero-project/monero/blob/6402dbee69867c76530b45e8a21b692850fea03f/src/cryptonote_basic/cryptonote_basic.h#L204). The code to serialize and deserialize objects of this class, e.g. for storing in the blockchain (binary) or for transmitting over RPC interfaces (JSON), isn't pretty and has to support a lot of cases, as can be seen [here](https://github.com/monero-project/monero/blob/6402dbee69867c76530b45e8a21b692850fea03f/src/cryptonote_basic/cryptonote_basic.h#L242) and [here](https://github.com/monero-project/monero/blob/6402dbee69867c76530b45e8a21b692850fea03f/src/ringct/rctTypes.h#L310). But the system holds so far.

With Seraphis we now have a new and very different transaction class `sp::SpTxSquashedV1`. Its current definition can be found [here](https://github.com/UkoeHB/monero/blob/3251219e160b7e82e8c6c7701b400df3148cefd4/src/seraphis/txtype_squashed_v1.h#L73).

To me it looks like the *single transaction class* approach breaks down with the advent of that Seraphis transaction class. The "overlap" of things that the two classes hold looks small, thus just adding everything from `SpTxSquashedV1` that does not yet have a counterpart in `transaction` to the latter class and afterwards eliminating the Seraphis class really does not look promising.

But is there a real problem at all in regard to coding? Can't we just double things like lists that now hold transactions, so there will be two lists, one holding the legacy transactions, and one holding Seraphis transactions?

I don't think this approach scales. For example if you have a daemon RPC call to return any transaction given its id, do we really want the returned struct to have two fields, for the two kinds of possible transaction classes? Or should calls to query the mempool really bother callers with two lists of transactions? How would you then go on and sort those into a single list ordered by reception time e.g. for display in a wallet?

And what happens in the future if Seraphis evolves? The version number `V1` in the class name `SpTxSquashedV1` hints at the general approach proposed by UkoeHB in his Seraphis code base to deal with more serious changes: Introduce `SpTxSquashedV2` which would then bring the number of transaction classe to 3 already. This has the potential to become a nightmare pretty quickly.

Currently I see only one possible solution: The introduction of some "generalized" or "unified" transaction class. An object of that class is able to alternatively hold a `cryptonote::transaction`, a `sp::SpTxSquashedV1`, or any future evolved Seraphis transaction object of course.

The technical means to make a suitable container for such disparate objects is a `boost::variant`. Read e.g. [this page](https://theboostcpplibraries.com/boost.variant) to see how those variants work. It seems that technically there is no other contender; using a C++ `union` construct directly surely isn't such, and why adding something "homegrown" to the Monero codebase if the Boost mechanism is already there.

And now an interesting historical tidbit that bolsters this proposal: It seems that back in 2012 or so the devs of the original CryptoNote codebase from which Monero is forked had in mind to implement several transaction types, or more exactly several types of inputs and outputs, with quite distinct properties. You can see this in code [here](https://github.com/monero-project/monero/blob/6402dbee69867c76530b45e8a21b692850fea03f/src/cryptonote_basic/cryptonote_basic.h#L154). And what do we have there? Exactly, `boost::variant`. (It does not matter that those additional types never saw any actual use.)

Of course there are many, many places in the Monero codebase that would be affected from replacing `cryptonote::transaction` with the new unified transaction class, so we have to consider this carefully, but I really see no other way to get a well-working system that includes Seraphis.

**Comment by rbrunner7** [comment]
One important technical question occured to me since I started this issue: If we replace `cryptonote::transaction` by something completely different in large parts of the codebase, how will binary deserialization still work?

The epee binary serialization that Monero uses is not "tagged", i.e. basically only field values get stored, but not their names. This way it's not possible, as far as I understand, to read / read ahead the first stored field value, look at its name, and then decide between "This is an old CryptoNote transaction" and "This is new variant based transaction".

But thankfully the very first field that gets stored for transactions now is a version number. Surprisingly for me, on the level of whole transactions, there are only 2 versions: Version 1 is pre RingCT, and Version 2 (the current version at the time of writing, see `CURRENT_TRANSACTION_VERSION`) is RingCT.

So any deserialization code can read the version number first; if it's 1 or 2, it can read a CryptoNote transaction and build a variant for it; if it's 3, it can read it as a variant, and depending on the variant type it will go on to deserialize a CrypoNote or a Seraphis transaction.

In short: That will probably work.

**Comment by ghost** [comment]
How about polymorphism rather than `variant`? Is there something that rules it out?
By the way, `variant` made its way into the standard library a while ago already (see https://en.cppreference.com/w/cpp/utility/variant).

As far as I can imagine, using variants will make every place that deals with transactions be handling the different types separately. Unless doing a kind of polymorphism through visitor lambdas...

I am thinking that using inheritance could force reasoning about modularizing transactions, rather than treating them as monolithic blocks. 

Even though different transaction classes may require some different handling, are they not fundamentally similar enough to be treated the same more often than not?

**Comment by rbrunner7** [comment]
Maybe you could give some rough sketches how those classes could look using polymorphism, using some rudimentary code / definition fragments? Right now I don't see yet where this idea could lead, and the fundamental differences are compared with using variants.

**Comment by UkoeHB** [comment]
@font-deny 
> How about polymorphism rather than variant? Is there something that rules it out?
> Even though different transaction classes may require some different handling, are they not fundamentally similar enough to be treated the same more often than not?

Transactions may have very different handling based on type, so type-erasure is the best bet.

Here are some guidelines about when to use different kinds of polymorphism:

- **Dynamic polymorphism (inheritance)**: when you have a fixed interface and want to slice away implementation details; derived classes should not need or want to add functionality.
- **Static polymorphism (templates)**: when you want to instantiate a generic set of behaviors for a local type (i.e. you write a template to encapsulate generic behaviors, then instantiate a template when you want those behaviors around your local type).
- **Mixed polymorphism (e.g. the Sean Parent pattern)**: when you have a generic set of behaviors that may be statically polymorphed but want a type-agnostic interface (i.e. your interface owns a handle to a sliced implementation that may be a templated derived class).
- **Type erasure (variants)**: when you have a general (not generic) pipeline and want to send multiple types through it, where handling of those types at exit points on the pipeline is type-specific. Note that using a pipeline implies the variant types have a 'same kind of thing' relationship.

In the case of Monero, type erasure is a good choice because there are many general workflows that apply to fixed sets of types that are known in advance, but which need to be independently handled at exit points on those workflows.

**Comment by DangerousFreedom1984** [comment]
After trying to analyse this problem I would like to share how I would approach it.

First, why do we need it at all? I can't fully answer as the places where the class `transaction` appears is pretty big, for example:

- blockchain_* > 10 files
- cryptonote_* > 5 files (definition)
- basically everywhere where the concept of transaction is used (wallet, pool, multisig, rpc_server, ...)

Since blocks are made of transactions and a blockchain is made of blocks, we need this class to write/read on the blockchain, which is where it has most of its use I guess.

Since this class is used in really a lot of places, I don't see any feasibility to modify every instance of this class to include the SpTxSquashedV1 (as also stated by rbrunner).

What would be the solution then?

Recreate the blockchain files, rpc_server, etc. And this is a huge task!

So, when the Seraphis hardfork happens, we will not only be using the seraphis_lib, which are in the folders seraphis_* (included the wallet) but also the legacy corresponding classes that would be in the folders seraphis_blockchain_db, seraphis_blockchain_utilities, seraphis_daemon, etc. What are your thoughts on that?

So to get started solving this problem, I believe I could create a variant that would use the old `transaction` class with the `SpTxSquashedV1` and if the transaction class scales, this approach is scalable as suggested. So initially I would only have some of the following [methods](https://github.com/monero-project/monero/blob/6402dbee69867c76530b45e8a21b692850fea03f/src/cryptonote_basic/cryptonote_basic.h#L226): 
```
    transaction();
    transaction(const transaction &t);
    transaction &operator=(const transaction &t);
    virtual ~transaction();
    void set_null();
    void invalidate_hashes();
    bool is_hash_valid() const { return hash_valid.load(std::memory_order_acquire); }
    void set_hash_valid(bool v) const { hash_valid.store(v,std::memory_order_release); }
    bool is_prunable_hash_valid() const { return prunable_hash_valid.load(std::memory_order_acquire); }
    void set_prunable_hash_valid(bool v) const { prunable_hash_valid.store(v,std::memory_order_release); }
    bool is_blob_size_valid() const { return blob_size_valid.load(std::memory_order_acquire); }
    void set_blob_size_valid(bool v) const { blob_size_valid.store(v,std::memory_order_release); }
    void set_hash(const crypto::hash &h) const { hash = h; set_hash_valid(true); }
    void set_prunable_hash(const crypto::hash &h) const { prunable_hash = h; set_prunable_hash_valid(true); }
    void set_blob_size(size_t sz) const { blob_size = sz; set_blob_size_valid(true); }
```

And I could make unit_tests to serialiaze/deserialize / test hash / test sizes and that would be all for now.

Doing more than that goes beyond my capacities and resources but the next steps would be to create the folders seraphis_blockchain_db, seraphis_blockchain_utilities, seraphis_daemon, etc, and use this class for its purpose. So it means that I obviously cant foresee everything that should be in that 'interface' yet but it would hopefully starting paving the way for the discussion on how txs should be written on the blockchain and how the queries from the blockchain should be done. What are your thoughts?

So to be a bit more precise on the tasks, I would create unit_tests to test similar functions to this one:

`bool is_hash_valid(const TransactionVariant &tx_variant)`

where `TransactionVariant = tools::variant<transaction, SpTxSquashedV1>`

and I would test with both a `transaction`  and a `SpTxSquashedV1`.

so if in the future we have something like SpTxSquashedV2, then it would be enough to add it to the Variant definition and handle the additional cases. What are your thoughts?

Well, it is basically the approach that @ShalitMonero was taking but I guess that there are some imprecisions in his code and I will try to 'go to the end'.

**Comment by Wolf54** [comment]
I'm new here, but I would like to share my opinion.

Looking at the code, from what I see, is that the entire transaction processing is manifested within the 'Cryptonote' classes.
That's ok as long as cryptonote is the only algorithm in town. If we add other ways to process a transaction, then we should split up the code into a base class and have separate classes for each way we process a transaction. After all, somebody decided very early on to write this in C++ and by doing so, this would be in the spirit of C++. Use polymorphism. Cryptonote would be the first specialized implementation for a transaction. SpTxSquashedV1 would be the next in line and others may follow.

The higher level code should have no knowledge of details on how a transaction is processed. The higher level code interacts only with the transaction base class which exposes the general ways a transaction is handled, like create, locate, store and load a transaction. Exactly how that is happening will be defined and executed by the specialized derived classes, like Pre-RingCT, Current-RingCT, SpTxSquashedV1 or may later SpTxSquashedVx. This has the advantage that all higher level code stays untouched whenever newer, better protocols get developed/added in the future. This will give great confidence that nothing in the existing code has changed if new code needs to be added. It would also make testing things easier. 

And yes, this is a huge task given the existing code which does not conform to C++ standards, but if we don't divide up the code into separate logical groups now, the pain will be even worse when the next better algorithm comes around. 

The problem with variants is, that sooner or later you always end up with lots and lots of if/then/else statements, spanning over hundreds of lines of code, making all of this hard to read and follow logical. That's the reason I would propose using C++ inheritance and make use of polymorphism offered by this language.

Here is an easy read with a good example:  https://www.geeksforgeeks.org/factory-method-for-designing-pattern
Just replace 'Vehicle' with 'Transaction' and you're almost there  :-)

What do you think?

**Comment by rbrunner7** [comment]
Thanks to @DangerousFreedom1984  and @Wolf54 to their recent contributions here.

I will try to write a summary what happened so far, which ideas and proposals are on the table, and where we stand with those.

As written documents so far we have this issue here and [this log](https://github.com/monero-project/meta/issues/809) of a Seraphis wallet workgroup meeting in March where we discussed this.

I think it's pretty easy to identify the following 3 principal approaches of dealing with the issue at hand that make sense with C++:

A) Polymorphism with a common transaction base class
B) Aggregation with a class that holds objects of all transaction classes as owned sub-objects
C) Variants with visitors

A) is what @Wolf54 described in their comment above and probably does not need much explanation: Just standard OO programming, using a common base class with a number of virtual methods, and as many derived classes as we need, probably 2 for now, for the CryptoNote transaction class, and the Seraphis transaction class, overriding those virtual methods, and of course adding class-specific methods as appropiate.

B) was discussed earlier and demonstrated in [this mock-up class](https://gist.github.com/ShalitMonero/3b40a5478f448f0d7efae9a5a7fe4a8e) written by @ShalitMonero. The principal transaction class is a brand-new class that holds / owns instances of the transactions of the various types, maybe privately, and offers a number of methods to work with them.

With C) here is no new class with code at the core, just a "naked" variant that can hold instances of all types of transactions. You deal with those using methods that take the variant as parameter and "visit" each possible transaction type; technically you can place those methods pretty much wherever you like.

@UkoeHB uses this approach already to quite some extent in the Seraphis library. An example: The legacy scanning code does not use `cryptonote::transaction` (as defined in the current Monero codebase [here](https://github.com/monero-project/monero/blob/master/src/cryptonote_basic/cryptonote_basic.h)) but 5 new classes for the 5 principal types of legacy transactions, and a variant to "bind then together" plus visitor based methods to process. You can see this in action [here](https://github.com/UkoeHB/monero/blob/seraphis_lib/src/seraphis_main/tx_component_types.h) and [here](https://github.com/UkoeHB/monero/blob/seraphis_lib/src/seraphis_main/tx_component_types.cpp)

**Comment by rbrunner7** [comment]
This is how I currently see these 3 approaches:

A) is probably what I would try first if I had to implement a system with 2 very different transaction classes given a chance to start out on a green field, with zero existing code. But looking at the base Monero code and the Seraphis library I think switching to this approach now is simply not feasible anymore. Sorry, @Wolf54 :)

B) could work, but is not very elegant, and maybe also error-prone as argued by @UkoeHB in the meeting I linked above e.g. because the compiler does not help you with processing all the types in all places.

C) looks most promising to me. @UkoeHB prefers this, obviously. @DangerousFreedom1984 also proposes this as the way to try in his recent comment.

I only see one technical problem with C) that I don't have a solution yet, and where I propose that somebody tries to find one soon: How to serialize and deserialize those "naked" variants in a way that is compatible with how it's done now, and compatible with the way dozens of gigabytes of transactions are stored in each of the thousands of copies of the Monero blockchain all over the world.

**Comment by jeffro256** [comment]
I propose the following approach:

1. Write two new classes which capture the information needed for old Cryptonote transactions: `CryptonoteTxV1` and `CryptonoteTxV2`
2. Write free functions for `CryptonoteTxV1`, `CryptonoteTxV2`, `SpTxSquashedV1`, and `SpTxCoinbaseV1` which [de]serialize each respective tx information and take two extra options: `pruned` and `versionless`. 
3. Make the main transaction variant `using MoneroTxVariant = tools::variant<CryptonoteTxV1, CryptonoteTxV2, SpTxSquashedV1, SpTxCoinbaseV1>`
4. The transaction versions for these transaction types will be 1, 2, 3, and 4, respectively.
5. The main [de]serialization function will first [de]serialize the version, then call the respective free function from step 2 through a generalized visitor. This helps us be compatible with the old transaction formats like @rbrunner7 was concerned about. When deserializing, we can value-initialize the variant based on the transaction version before using the visitor. See this commit for an idea of what I mean: https://github.com/jeffro256/monero/commit/2e0f9b3b1a351a7dd5362b9fe0a00f6ff05855bd. This will not take extra `if` branches or any extra code when adding transaction classes to the variant if we do it correctly like @Wolf54 was worried about.
6. If we don't want to modify the existing consensus code, perhaps in order to minimize silly transcribing mistakes, we can write functions which map `CryptonoteTxV1` and `CryptonoteTxV2` into an instance of `cryptonote::transaction` which represents the same information. Then in order to verify `MoneroTxVariant` for V1 and V2 transactions, we transform it into a `cryptonote::transaction` and then verify that. Then there's only one new function we need to check to prevent old consensus rule mistakes, and we only need to check that it correctly transforms our transaction classes into a `cryptonote::transaction`. 

There are two major reasons why we might want to write new transaction classes for Cryptonote transactions instead of reusing `cryptonote::transaction`:

a. `cryptonote::transaction` contains caches for the hash, prunable hash, and blob size. We will most likely want our new transaction class to also have these (and perhaps other) properties cached. Making our caches for our all-encompassing tx mesh with these fields will lead to messy code, and those copying/moving those values will take up CPU cycles even when not using them.
b. To keep in good programming practices which align with the design principles of Seraphis lib, we should try to minimize using "everything" classes which have more fields than they need to perform their purpose. For example: `cryptonote::transaction` has two fields `signatures` and `rct_signatures`, even though only one is ever used at a time. Another example: not all RingCT fields in types in files `rctTypes.*` are serialized in a transaction, yet they are fields in `cryptonote::transaction`. This leads to messy and fragile logic like in `Blockchain::expand_transaction_2()` which takes a `cryptonote::transaction` *mutable* reference, and populates RingCT fields in the tx like `mixRing`. Later, we take a reference to the tx's `rct_signatures` field and validate against that.

**Comment by rbrunner7** [comment]
Thanks, @jeffro256, for this already quite detailed proposal!

> This helps us be compatible with the old transaction formats like @rbrunner7 was concerned about.

I am not sure from what you write whether this scheme is able to achieve full "deserializing compatibility". I mean with this that the new code supporting current as well as Seraphis transactions will be able to read current blockchain files *as is*, i.e. without a conversion step that is necessary in order to start using the new software.

I think a requirement to go through a conversion first with current blockchain files would be quite problematic for updating. If you need an amount of free space that is similar to the size of the blockchain to convert, conversion will be impossible for many people simply for lack of disk space. And even if you can convert "in place" without needing much additional space, this may result in conversion run times of hours for many people.

Maybe 6 instead of only 4 numbers would be needed for such a full compatibility approach? Version 1 and 2 current transactions in their current format, 3 and 4 for `CryptonoteTxV1` and `CryptonoteTxV2`, and 5 and 6 for the Seraphis types?

**Comment by rbrunner7** [comment]
@jeffro256, did you already try to get some overview how much code is in need of extensive rework if the new main transaction class switches from `cryptonote::transaction` to a completely different `MoneroTxVariant`? While your approach is certainly very clean I worry a bit about the amount of work needed to implement it.

For the sake of completeness, I shortly describe one more approach that came up in one of the last workgroup meetings that I think was not yet mentioned in this issue here, an approach that borders on a hack but might be quite an easy way out i.e. may need only a minimum of refactoring of existing code, mostly in the daemon:

Keep `cryptonote::transaction` and make only a single small modification to it: Add one new field to hold a variant object for Seraphis transactions. If the fields is null, it's an "old" transaction, and if the field is filled, it's a Seraphis transaction, and many of the other fields probably are not needed  for processing.

**Comment by jeffro256** [comment]
> I am not sure from what you write whether this scheme is able to achieve full "deserializing compatibility". I mean with this that the new code supporting current as well as Seraphis transactions will be able to read current blockchain files as is, i.e. without a conversion step that is necessary in order to start using the new software.

Yes it will be. 

> Version 1 and 2 current transactions in their current format, 3 and 4 for CryptonoteTxV1 and CryptonoteTxV2, and 5 and 6 for the Seraphis types?

`CryptonoteTxV1` and `CryptonoteTxV2` *would* be version 1 and 2 transactions, just under a different class. I wouldn't be adding any new transaction formats.

**Comment by DangerousFreedom1984** [comment]
Hello jeffro,

thank you for looking at this issue! I think it is great to use this transaction interface as you propose. To better understand your proposal, let's try to make a step ahead and see how your structure would be used.

So, I guess my questions would be:
1) Where do we put these free functions and the new Variant tx class? How do you see it being structured from the folder organization point of view?

2) how to make a link between the necessary new [database scheme](https://github.com/UkoeHB/monero/blob/seraphis_lib/src/blockchain_db/lmdb/db_lmdb.cpp#L182) that we need to build for Seraphis and the [mock ledger](https://github.com/UkoeHB/monero/blob/seraphis_lib/src/seraphis_mocks/mock_ledger_context.h) used for testing?

3) how hard would it be to:
    1) Create the tables (db entries) with the necessary tx info from Seraphis
    2) Initialize a new DB object
    3) Save txs in this database
    4) Load txs from this database to update wallet status

**Comment by jeffro256** [comment]
I've started writing the code and it is moving very quickly IMO. Here's a commit from tonight: https://github.com/jeffro256/monero/commit/dae6b66203b56c6cc154fdfa8250cfe5b0f229b7#diff-f8e11e865916f4f2e5b34c8c36e06bc25d6ffba1800ed0204bb50768be3ece73

To answer your question one, @DangerousFreedom1984, I created some RingCT-related types and put in a file in the seraphis_core folder. I then created four new transactions types for v1, v1 coinbase, v2, and v2 coinbase, and put those in seraphis_main folder by `SpTxSquashedV1` and `SpTxCoinbaseV1`. I then wrote up the serialization code and stuffed it into files called cryptonote_tx_serialization in the seraphis_impl folder.

As for question 2, we won't have to touch the schema IIRC. In other words, we won't need an LMDB migration. But we need to make sure that the `BlockchainLMDB` interface understands `MoneroTxVariant` instead of `cryptonote::transaction` in the future.

As for question 3, that's basically what we're going to need to core / blockchain validations rules for. We need to update the transaction rules, and then store the Seraphis tx blobs into the database (and be able to read them). This will take a lot of effort and review work to make sure we get it right.

However, if you look in that commit, I've written a unit test which tests that v2 RingCT transactions can be loaded correctly, which is honestly the hardest part. Now all the remains is loading coinbase and v1 transactions, and then storing (storing is easier than loading).

**Comment by vtnerd** [comment]
> As for question 2, we won't have to touch the schema IIRC. In other words, we won't need an LMDB migration. But we need to make sure that the BlockchainLMDB interface understands MoneroTxVariant instead of cryptonote::transaction in the future.

Why the variant? Was there a discussion elsewhere (#no-wallet-left-behind ? )  It's probably going to be a pain, even with all the helper classes.

**Comment by jeffro256** [comment]
Yeah there was some discussion in that room over the last few months. In my opinion, it will be less of a pain than having everything in one class like we currently do, especially if we plan on moving to a different EC, or implementing full-chain membership proofs, or other transaction formats we can't foresee yet. Breaking down the different types of valid transactions into non-malleable, concrete types makes transaction verification much simpler in my opinion, which is the most important part to get right. The alternative is to stuff everything into `cryptonote::transaction` or use polymorphism, which have their own downsides that outweigh the downsides of variants for this use case, IMO.

**Comment by jeffro256** [comment]
For example, take this class declaration (which is part of the `MoneroTxVariant`):

https://github.com/jeffro256/monero/blob/aad86a91617ced128852d92bca78643789f61f5b/src/seraphis_main/txtype_cryptonote_coinbase_v1.h#L42-L47

The only fields we care about is height, vout, and extra. We only have to verify 3 fields! We don't have to check that RCT type == 0, unlock time == height + 60, vin is size 1, vin[0] is txin_gen type, all vout types are txout_to_key, signatues is empty, seraphis member proofs is empty, serpahis composition proofs is empty, tx supplement is empty, seraphis linking tags is empty, discretized fee is 0, etc, etc. We don't have to check this because that information isn't possible to be represented. So we only have to check that the deserialization process is non-malleable, and then we get to forget about the rest of all possible information that could show up in a transaction. In my opinion, this makes the code way, way, way more simple to read and review correctness, even if it takes more writing up front.
